### PR TITLE
New Post

### PR DESCRIPTION
**Pull request by jameskerr**

### CODE REVIEW

File: content/posts/partially-controlled-react-components.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+  state: {
+    columnVisibility, // visiblity value
+    columnOrder, // order value
+  },
+  onColumnVisibilityChange: setColumnVisibility, // visibility setter
+  onColumnOrderChange: setColumnOrder, // order setter
+});
+```
+
+Take a look at that partial control! I really think he's onto something. Since the component/hook does so many things, the state is divided into logical slices. Each slice can either be controlled or uncontrolled depending on the presence of an `on[SliceName]Change` handler. I love this pattern. However, to me, its undesirable that the value and change handler are not co-located. It is also a bit verbose. But it is declarative and much more "React".
```
Comment (jamiesunderland) [inline content/posts/partially-controlled-react-components.md]: another downside to point out about this pattern (though a bit of an edge case), is the inability to short circuit safely when dealing with async state. If for example you need some asynchronous data that must be provided to render the component in the hook, you now need to wrap the calling component inside another component to short circuit.

for example image:

Component version:
```jsx
const userQuery = useUserQuery();

return (
  <div>
    {!userQuery.data?.user && (
      <Spinner/>
    )}
    {userQuery.data?.user && (
      <UserTable user={userQuery.data.user}/>
    )}
  </div>
);

```

Might not work as:
```jsx
const userQuery = useUserQuery();
const userTable = useUserTable({user: userQuery.data?.user});

return (
  <div>
    {!userQuery.data?.user && (
      <Spinner/>
    )}
    {userQuery.data?.user && userTable}
  </div>
);

```

File: content/posts/partially-controlled-react-components.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    value: editing
+    onChange: (state) => setEditing(state)
+  }}
+  expandedState={{
+    defaultValue: expanded
+  }}
+/>
+```
+
+Each logical group of state has it's own prop object for user's to control. The presence of this prop indicates to the component that its state slice will be managed externally.
```
Comment (jamiesunderland) [inline content/posts/partially-controlled-react-components.md]: I do like this. Another pattern I've seen (especially with drag and drop libraries) and enjoyed is the hook controls pattern. So you do something like

```jsx
  const treeControls = useTreeControls({
    selectionState: {
      value: selection,
      onChange: (state) => setSelection(state)
    },
    editingState: {
      value: editing
      onChange: (state) => setEditing(state)
    },
    expandedState: {
      defaultValue: expanded
    }
  })
  return (
    <Tree controls={treeControls}/>
  );
```

It's the same idea but avoids over-proping and makes memoization a little easier on the eyes.

**Review by jamiesunderland — APPROVED** [review]

File: content/posts/partially-controlled-react-components.md (hunk 1 above)
Comment (jameskerr) [reply content/posts/partially-controlled-react-components.md]: Good point.

File: content/posts/partially-controlled-react-components.md (hunk 2 above)
Comment (jameskerr) [reply content/posts/partially-controlled-react-components.md]: Yes true!

Comment (SevenOutman) [reply content/posts/partially-controlled-react-components.md]: Much agreed. Component with props are best for reusing UI, but hooks are the true answer to reusing logic.

BTW it kinda reminds me of dataSource/delegate patterns in iOS development.

**Pull request merged**